magda-correspondence-api/src/CEmailTplRender.ts

```diff
@@ -0,0 +1,82 @@
+import { Attachment } from "./SMTPMailer";
+import CContentApiDirMapper from "./CContentApiDirMapper";
+import * as path from "path";
+import * as mimeTypes from "mime-types";
+import * as Mustache from "mustache";
+
+/**
+ * customised mustache render function.
+ * usgae: {{#inlineImg}}<img class="xxx" src="emailTpls/assets/1.png" />{{/inlineImg}}
+ *
+ * @param text
+ * @param render
+ */
+function inlineImg(
+    this: CEmailTplRender,
+    text: string,
+    render: (text: string) => string
+) {
+    // --- render inner block first
+    const content = render(text);
+    const matches = content.match(/<img\s+[^>]*src=["']([^"']*)['"]/i);
+    if (!matches) {
+        return content;
+    }
+    const filePath = matches[1];
+    const fileName = path.basename(filePath);
+    const idx = this.attachments.findIndex(item => item.path === filePath);
+    let cid = "";
+    if (idx !== -1) {
+        cid = this.attachments[idx].cid;
+    } else {
+        const mimeType = mimeTypes.lookup(filePath);
+        cid = `inlineImg_${Math.random()}`.replace(".", "_");
+        this.attachments.push({
+            filename: fileName,
+            contentType:
+                mimeType === false ? "application/octet-stream" : mimeType,
+            contentDisposition: "inline",
+            path: filePath,
+            cid
+        });
+    }
+    return content.replace(
+        /(<img\s+[^>]*src=["'])([^"']*)(['"])/i,
+        `$1${"cid:" + cid}$3`
+    );
+}
+
+export default class CEmailTplRender {
```

Comment:
Just a pet peeve but could we please use descriptive names? E.g. Template instead of Tpl.

Reply:
I think it's a good idea~
Have updated the name to `CEmailTemplateRender`

Reply:
Thanks. Did a quick check and there are a few other instances of *Tpls* as well.

Reply:
Removed Tpls
And move assets to `emailTemplates` 👍